distributed_shampoo/fsdp_shampoo.py

```diff
@@ -363,33 +372,86 @@ def _initialize_preconditioners_and_steps(self):
-                    use_dtensor=self._use_dtensor,
-                )
+                if self._convex_shape_recovery == ConvexShapeRecoveryMethod.SPLIT:
+                    state[PRECONDITIONERS] = SplitShampooPreconditioner(
+                        p,
+                        self._param_metadata[p],
+                        large_dim_method=self._large_dim_method,
+                        beta2=group[BETAS][1],
+                        epsilon=group[EPSILON],
+                        exponent_override=self._exponent_override,
+                        exponent_multiplier=self._exponent_multiplier,
+                        use_bias_correction=self._use_bias_correction,
+                        max_preconditioner_dim=self._max_preconditioner_dim,
+                        dtype=self._preconditioner_dtype,
+                        idx=idx,
+                        use_merge_dims=self._use_merge_dims,
+                        start_preconditioning_step=self._start_preconditioning_step,
+                        grafting_type=self._grafting_type,
+                        grafting_beta2=self._grafting_beta2,
+                        grafting_epsilon=self._grafting_epsilon,
+                        use_protected_eigh=self._use_protected_eigh,
+                        use_dtensor=self._use_dtensor,
+                    )
+                elif self._convex_shape_recovery == ConvexShapeRecoveryMethod.COMM:
+                    state[PRECONDITIONERS] = CommunicationShampooPreconditioner(
+                        p,
+                        self._param_metadata[p],
+                        large_dim_method=self._large_dim_method,
+                        beta2=group[BETAS][1],
+                        epsilon=group[EPSILON],
+                        exponent_override=self._exponent_override,
+                        exponent_multiplier=self._exponent_multiplier,
+                        use_bias_correction=self._use_bias_correction,
+                        max_preconditioner_dim=self._max_preconditioner_dim,
+                        dtype=self._preconditioner_dtype,
+                        idx=idx,
+                        use_merge_dims=self._use_merge_dims,
+                        start_preconditioning_step=self._start_preconditioning_step,
+                        grafting_type=self._grafting_type,
+                        grafting_beta2=self._grafting_beta2,
+                        grafting_epsilon=self._grafting_epsilon,
+                        use_protected_eigh=self._use_protected_eigh,
+                        use_dtensor=self._use_dtensor,
+                        # TODO: can do more complex assignment here
+                        left_comm=CommunicationType.NONE
+                        if group_rank == 0
+                        else CommunicationType.RECV,
+                        right_comm=CommunicationType.NONE
+                        if group_rank == dist.get_world_size() - 1
+                        else CommunicationType.SEND,
```

Comment:
Yes, I was thinking of abstracting these into a function or at least defining them outside of the preconditioner initialization when we enable different strategies! `CommunicationDirection` is actually not related to this - the only purpose of `CommunicationDirection` is so that I don't need to create separate functions for communicating forward (at the beginning of the step) and communicating backward (towards the end of the step) 😛

Reply:
I see. Is `CommunicationDirection` something that could be a boolean variable instead then?